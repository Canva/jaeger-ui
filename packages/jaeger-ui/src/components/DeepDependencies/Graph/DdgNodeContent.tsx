// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import cx from 'classnames';
import { TMeasureNodeUtils, TRendererUtils } from '@jaegertracing/plexus/lib/Digraph/types';
import { TLayoutVertex, TVertex } from '@jaegertracing/plexus/lib/types';

import { focalNodeIcon, setFocusIcon } from './node-icons';
import { getUrl } from '../url';
import NewWindowIcon from '../../common/NewWindowIcon';
import { getUrl as getSearchUrl } from '../../SearchTracePage/url';
import { EViewModifier, TDdgVertex, PathElem } from '../../../model/ddg/types';

import './DdgNodeContent.css';

type TProps = {
  focalNodeUrl: string | null;
  getVisiblePathElems: (vertexKey: string) => PathElem[] | undefined;
  isFocalNode: boolean;
  isPositioned: boolean;
  operation: string | null;
  service: string;
  setViewModifier: (vertexKey: string, viewModifier: EViewModifier, isEnabled: boolean) => void;
  vertexKey: string;
};

// While browsers suport URLs of unlimited length, many server clients do not handle more than this max
const MAX_LENGTH = 2083;
const MAX_LINKED_TRACES = 35;
const MIN_LENGTH = getSearchUrl().length;
const PARAM_NAME_LENGTH = '&traceID='.length;

export default class DdgNodeContent extends React.PureComponent<TProps> {
  static measureNode(_: TVertex<any>, utils: TMeasureNodeUtils) {
    const { height, width } = utils.getWrapperSize();
    return {
      height: height + 2,
      width: width + 2,
    };
  }

  static getNodeRenderer(
    getVisiblePathElems: (vertexKey: string) => PathElem[] | undefined,
    setViewModifier: (vertexKey: string, viewModifier: EViewModifier, enable: boolean) => void
  ) {
    return function renderNode(vertex: TDdgVertex, utils: TRendererUtils, lv: TLayoutVertex<any> | null) {
      const { isFocalNode, key, operation, service } = vertex;
      return (
        <DdgNodeContent
          focalNodeUrl={isFocalNode ? null : getUrl({ operation, service })}
          getVisiblePathElems={getVisiblePathElems}
          isFocalNode={isFocalNode}
          isPositioned={Boolean(lv)}
          operation={operation}
          setViewModifier={setViewModifier}
          service={service}
          vertexKey={key}
        />
      );
    };
  }

  private viewTraces = () => {
    const { vertexKey, getVisiblePathElems } = this.props;
    const elems = getVisiblePathElems(vertexKey);
    if (elems) {
      const ids: Set<string> = new Set();
      let currLength = MIN_LENGTH;
      for (let i = 0; i < elems.length; i++) {
        const id = elems[i].memberOf.traceID;
        if (ids.has(id)) {
          continue;
        }
        // Keep track of the length, then break if it is too long, to avoid opening a tab with a URL that the
        // backend cannot process, even if there are more traceIDs
        currLength += PARAM_NAME_LENGTH + id.length;
        if (currLength > MAX_LENGTH) {
          break;
        }
        ids.add(id);
        if (ids.size >= MAX_LINKED_TRACES) {
          break;
        }
      }
      window.open(getSearchUrl({ traceID: Array.from(ids) }), '_blank');
    }
  };

  private onMouseUx = (event: React.MouseEvent<HTMLElement>) => {
    const { vertexKey, setViewModifier } = this.props;
    setViewModifier(vertexKey, EViewModifier.Hovered, event.type === 'mouseover');
  };

  render() {
    const { focalNodeUrl, isFocalNode, isPositioned, operation, service } = this.props;
    return (
      <div className="DdgNodeContent" onMouseOver={this.onMouseUx} onMouseOut={this.onMouseUx}>
        <div
          className={cx('DdgNodeContent--core', {
            'is-focalNode': isFocalNode,
            'is-positioned': isPositioned,
          })}
        >
          {isFocalNode && <div className="DdgNodeContent--focalMarker">{focalNodeIcon}</div>}
          <div>
            <h4 className="DdgNodeContent--label">{service}</h4>
            {operation && <div className="DdgNodeContent--label">{operation}</div>}
          </div>
        </div>

        <div className="DdgNodeContent--actionsWrapper">
          {focalNodeUrl && (
            <a href={focalNodeUrl} className="DdgNodeContent--actionsItem">
              {setFocusIcon}
              <span className="DdgNodeContent--actionsItemText">Set focus</span>
            </a>
          )}
          <a className="DdgNodeContent--actionsItem" onClick={this.viewTraces} role="button">
            <NewWindowIcon />
            <span className="DdgNodeContent--actionsItemText">View traces</span>
          </a>
        </div>
      </div>
    );
  }
}
