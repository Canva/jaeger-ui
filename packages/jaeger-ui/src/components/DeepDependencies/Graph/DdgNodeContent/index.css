/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.DdgNodeContent--core {
  background: #eee;
  border-radius: 100%;
  box-shadow: 0 1px 2px 1px rgba(0, 0, 0, 0.15);
  cursor: default;
  display: flex;
}

.DdgNodeContent--core.is-positioned {
  bottom: 1px;
  left: 1px;
  position: absolute;
  right: 1px;
  top: 1px;
}


.DdgNodeContent--core.is-focalNode {
  /* background: linear-gradient(90deg, #eb2f96 0%, #eb2f96 14%, #eee 14%); */
}

.DdgNodeContent--core.is-focalNode::before {
  /* background-image: repeating-linear-gradient(-60deg, #fffb8f, #fff566 3px, #fffb8f 5px); */
  border: 1px dashed #eb2f96;
  border-radius: 100%;
  bottom: -20px;
  content: '';
  left: -20px;
  position: absolute;
  right: -20px;
  top: -20px;
  z-index: -3;
  margin: 1px 0 0 1px;
}
/*
.DdgNodeContent--core.is-focalNode::before {
  bottom: 0;
  content: '';
  left: 0;
  outline: 20px solid rgba(255, 255, 255, 0.2);
  position: absolute;
  right: 0;
  top: 0;
}

.DdgNodeContent--core.is-focalNode::after {
  bottom: 0;
  content: '';
  left: 0;
  outline-offset: 20px;
  outline: 1px dashed #ec8cb475;
  position: absolute;
  right: 0;
  top: 0;
}

.DdgNodeContent--focalMarker {
  align-items: center;
  background: #eb2f96;
  display: flex;
  padding: 0.3rem 0.15rem;
}
*/

.DdgNodeContent--label {
  margin: 0;
  /* padding: 0.5rem 0.75rem 0.3rem 0.75rem; test fit */
  margin-right: auto;
  margin-left: auto;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
}

.DdgNode--test-classname {
  /* margin: auto; */
  margin-right: auto;
  margin-left: auto;
}

.DdgNodeContent--label:only-child {
  /* padding: 0.5rem; test fit */
}

.DdgNodeContent--label:not(:first-child) {
  /* padding-top: 0.2rem; test fit */
  padding-top: 5px;
}

.DdgNodeContent--actionsWrapper {
  background: #fff;
  border-radius: 3px;
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
  bottom: 100%;
  box-shadow: 0 0px 4px 1px rgba(0, 0, 0, 0.1);
  left: 1em;
  opacity: 0;
  pointer-events: none;
  position: absolute;
  transition: opacity 0.1s;
  white-space: nowrap;
}

.DdgNodeContent:hover > .DdgNodeContent--actionsWrapper {
  opacity: 1;
  pointer-events: all;
}

.DdgNodeContent--actionsItem {
  align-items: center;
  display: flex;
  padding: 0.5em;
}

.DdgNodeContent--actionsItem:not(:hover) {
  color: inherit;
}

.DdgNodeContent--actionsItemText {
  font-size: 0.9em;
  line-height: normal;
  margin-left: 0.5em;
}
