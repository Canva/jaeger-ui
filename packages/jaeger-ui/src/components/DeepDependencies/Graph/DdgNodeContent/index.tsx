// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Checkbox, Popover } from 'antd';
import cx from 'classnames';
import { TLayoutVertex } from '@jaegertracing/plexus/lib/types';
import IoAndroidLocate from 'react-icons/lib/io/android-locate';
import MdVisibilityOff from 'react-icons/lib/md/visibility-off';

import calcPositioning from './calc-positioning';
import {
  MAX_LENGTH,
  MAX_LINKED_TRACES,
  MIN_LENGTH,
  OP_PADDING_TOP,
  PARAM_NAME_LENGTH,
  RADIUS,
  WORD_RX,
} from './constants';
import { setFocusIcon } from './node-icons';
import { trackSetFocus, trackViewTraces } from '../../index.track';
import { getUrl } from '../../url';
import BreakableText from '../../../common/BreakableText';
import FilteredList from '../../../common/FilteredList';
import NewWindowIcon from '../../../common/NewWindowIcon';
import { getUrl as getSearchUrl } from '../../../SearchTracePage/url';
import {
  ECheckedStatus,
  EDdgDensity,
  EDirection,
  EViewModifier,
  TDdgVertex,
  PathElem,
} from '../../../../model/ddg/types';

import './index.css';

type TProps = {
  focalNodeUrl: string | null;
  focusPathsThroughVertex: (vertexKey: string) => void;
  getGenerationVisibility: (vertexKey: string, direction: EDirection) => ECheckedStatus | null;
  getVisiblePathElems: (vertexKey: string) => PathElem[] | undefined;
  hideVertex: (vertexKey: string) => void;
  isFocalNode: boolean;
  isPositioned: boolean;
  operation: string | string[] | null;
  service: string;
  setOperation: (operation: string) => void;
  setViewModifier: (vertexKey: string, viewModifier: EViewModifier, isEnabled: boolean) => void;
  updateGenerationVisibility: (vertexKey: string, direction: EDirection) => void;
  vertexKey: string;
};

export default class DdgNodeContent extends React.PureComponent<TProps> {
  hoverClearDelay?: number;

  state = {
    hovered: false,
  };

  static measureNode() {
    const diameter = 2 * (RADIUS + 1);

    return {
      height: diameter,
      width: diameter,
    };
  }

  static getNodeRenderer({
    baseUrl,
    density,
    extraUrlArgs,
    focusPathsThroughVertex,
    getGenerationVisibility,
    getVisiblePathElems,
    hideVertex,
    setOperation,
    setViewModifier,
    updateGenerationVisibility,
  }: {
    baseUrl: string;
    density: EDdgDensity;
    extraUrlArgs?: { [key: string]: unknown };
    focusPathsThroughVertex: (vertexKey: string) => void;
    getGenerationVisibility: (vertexKey: string, direction: EDirection) => ECheckedStatus | null;
    getVisiblePathElems: (vertexKey: string) => PathElem[] | undefined;
    hideVertex: (vertexKey: string) => void;
    setOperation: (operation: string) => void;
    setViewModifier: (vertexKey: string, viewModifier: EViewModifier, enable: boolean) => void;
    updateGenerationVisibility: (vertexKey: string, direction: EDirection) => void;
  }) {
    return function renderNode(vertex: TDdgVertex, _: unknown, lv: TLayoutVertex<any> | null) {
      const { isFocalNode, key, operation, service } = vertex;
      return (
        <DdgNodeContent
          focalNodeUrl={
            isFocalNode ? null : getUrl({ density, operation, service, ...extraUrlArgs }, baseUrl)
          }
          focusPathsThroughVertex={focusPathsThroughVertex}
          getGenerationVisibility={getGenerationVisibility}
          getVisiblePathElems={getVisiblePathElems}
          hideVertex={hideVertex}
          isFocalNode={isFocalNode}
          isPositioned={Boolean(lv)}
          operation={operation}
          setOperation={setOperation}
          setViewModifier={setViewModifier}
          service={service}
          updateGenerationVisibility={updateGenerationVisibility}
          vertexKey={key}
        />
      );
    };
  }

  componentWillUnmount() {
    if (this.state.hovered) {
      this.props.setViewModifier(this.props.vertexKey, EViewModifier.Hovered, false);
    }
  }

  private focusPaths = () => {
    const { focusPathsThroughVertex, vertexKey } = this.props;
    focusPathsThroughVertex(vertexKey);
  };

  private hideVertex = () => {
    const { hideVertex, vertexKey } = this.props;
    hideVertex(vertexKey);
  };

  private updateChildren = () => {
    const { updateGenerationVisibility, vertexKey } = this.props;
    updateGenerationVisibility(vertexKey, EDirection.Downstream);
  };

  private updateParents = () => {
    const { updateGenerationVisibility, vertexKey } = this.props;
    updateGenerationVisibility(vertexKey, EDirection.Upstream);
  };

  private viewTraces = () => {
    trackViewTraces();
    const { vertexKey, getVisiblePathElems } = this.props;
    const elems = getVisiblePathElems(vertexKey);
    if (elems) {
      const urlIds: Set<string> = new Set();
      let currLength = MIN_LENGTH;
      // Because there is a limit on traceIDs, attempt to get some from each elem rather than all from one.
      const allIDs = elems.map(({ memberOf }) => memberOf.traceIDs.slice());
      while (allIDs.length) {
        const ids = allIDs.shift();
        if (ids && ids.length) {
          const id = ids.pop();
          if (id && !urlIds.has(id)) {
            // Keep track of the length, then break if it is too long, to avoid opening a tab with a URL that
            // the backend cannot process, even if there are more traceIDs
            currLength += PARAM_NAME_LENGTH + id.length;
            if (currLength > MAX_LENGTH) {
              break;
            }
            urlIds.add(id);
            if (urlIds.size >= MAX_LINKED_TRACES) {
              break;
            }
          }
          allIDs.push(ids);
        }
      }
      window.open(getSearchUrl({ traceID: Array.from(urlIds) }), '_blank');
    }
  };

  private onMouseUx = (event: React.MouseEvent<HTMLElement>) => {
    const { vertexKey, setViewModifier } = this.props;
    const hovered = event.type === 'mouseenter';
    setViewModifier(vertexKey, EViewModifier.Hovered, hovered);
    if (hovered) {
      if (this.hoverClearDelay) {
        clearTimeout(this.hoverClearDelay);
        this.hoverClearDelay = undefined;
      } else {
        this.setState({ hovered });
      }
    } else {
      this.hoverClearDelay = setTimeout(() => {
        this.setState({ hovered });
        this.hoverClearDelay = undefined;
      }, 150);
    }
  };

  render() {
    const { hovered } = this.state;
    const {
      focalNodeUrl,
      getGenerationVisibility,
      isFocalNode,
      isPositioned,
      operation,
      service,
      setOperation,
      vertexKey,
    } = this.props;

    const { radius, svcWidth, opWidth, svcMarginTop } = calcPositioning(service, operation);
    const scaleFactor = RADIUS / radius;
    const transform = `translate(${RADIUS - radius}px, ${RADIUS - radius}px) scale(${scaleFactor})`;

    const childrenVisibility = hovered && getGenerationVisibility(vertexKey, EDirection.Downstream);
    const parentVisibility = hovered && getGenerationVisibility(vertexKey, EDirection.Upstream);

    return (
      <div className="DdgNodeContent" onMouseEnter={this.onMouseUx} onMouseLeave={this.onMouseUx}>
        <div
          className={cx('DdgNodeContent--core', {
            'is-focalNode': isFocalNode,
            'is-positioned': isPositioned,
          })}
          style={{ width: `${radius * 2}px`, height: `${radius * 2}px`, transform }}
        >
          <div className="DdgNodeContent--labelWrapper">
            <h4
              className="DdgNodeContent--label"
              style={{ marginTop: `${svcMarginTop}px`, width: `${svcWidth}px` }}
            >
              <BreakableText text={service} wordRegexp={WORD_RX} />
            </h4>
            {operation && (
              <div
                className="DdgNodeContent--label"
                style={{ paddingTop: `${OP_PADDING_TOP}px`, width: `${opWidth}px` }}
              >
              {
                Array.isArray(operation)
                  ? <Popover
                      content={
                        <FilteredList
                          cancel={() => {}}
                          options={operation}
                          value={null}
                          setValue={setOperation}
                        />
                      }
                      placement="bottom"
                      title="Select Operation to Filter Graph"
                    >
                      <span>{`${operation.length} Operations`}</span>
                    </Popover>
                  : <BreakableText text={operation} wordRegexp={WORD_RX} />
              }
              </div>
            )}
          </div>
        </div>

        {hovered && (
          <div className="DdgNodeContent--actionsWrapper">
            {focalNodeUrl && (
              <a href={focalNodeUrl} className="DdgNodeContent--actionsItem" onClick={trackSetFocus}>
                <span className="DdgNodeContent--actionsItemIconWrapper">{setFocusIcon}</span>
                <span className="DdgNodeContent--actionsItemText">Set focus</span>
              </a>
            )}
            <a className="DdgNodeContent--actionsItem" onClick={this.viewTraces} role="button">
              <span className="DdgNodeContent--actionsItemIconWrapper">
                <NewWindowIcon />
              </span>
              <span className="DdgNodeContent--actionsItemText">View traces</span>
            </a>
            {!isFocalNode && (
              <a className="DdgNodeContent--actionsItem" onClick={this.focusPaths} role="button">
                <span className="DdgNodeContent--actionsItemIconWrapper">
                  <IoAndroidLocate />
                </span>
                <span className="DdgNodeContent--actionsItemText">Focus paths through this node</span>
              </a>
            )}
            {!isFocalNode && (
              <a className="DdgNodeContent--actionsItem" onClick={this.hideVertex} role="button">
                <span className="DdgNodeContent--actionsItemIconWrapper">
                  <MdVisibilityOff />
                </span>
                <span className="DdgNodeContent--actionsItemText">Hide node</span>
              </a>
            )}
            {parentVisibility && (
              <a className="DdgNodeContent--actionsItem" onClick={this.updateParents} role="button">
                <span className="DdgNodeContent--actionsItemIconWrapper">
                  <Checkbox
                    checked={parentVisibility === ECheckedStatus.Full}
                    indeterminate={parentVisibility === ECheckedStatus.Partial}
                  />
                </span>
                <span className="DdgNodeContent--actionsItemText">View Parents</span>
              </a>
            )}
            {childrenVisibility && (
              <a className="DdgNodeContent--actionsItem" onClick={this.updateChildren} role="button">
                <span className="DdgNodeContent--actionsItemIconWrapper">
                  <Checkbox
                    checked={childrenVisibility === ECheckedStatus.Full}
                    indeterminate={childrenVisibility === ECheckedStatus.Partial}
                  />
                </span>
                <span className="DdgNodeContent--actionsItemText">View Children</span>
              </a>
            )}
          </div>
        )}
      </div>
    );
  }
}
