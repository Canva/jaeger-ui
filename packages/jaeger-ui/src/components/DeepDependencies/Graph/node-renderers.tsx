// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import cx from 'classnames';
import { TLayoutVertex } from '@jaegertracing/plexus/lib/types';

import { TDdgVertex } from '../../../model/ddg/types';

import './node-renderers.css';

// eslint-disable-next-line import/prefer-default-export
export function getFindEmphasisRenderers(findMatches: Set<TDdgVertex>) {
  function renderVectorBorder(lv: TLayoutVertex<TDdgVertex>) {
    const className = cx('DdgNode-VectorBorder', {
      'is-findMatch': findMatches.has(lv.vertex),
      'is-focalNode': lv.vertex.isFocalNode,
    });
    return (
      <rect
        className={className}
        vectorEffect="non-scaling-stroke"
        width={lv.width - 2}
        height={lv.height - 2}
        x="1"
        y="1"
      />
    );
  }

  function renderVectorFindEmphasisOutline(lv: TLayoutVertex<any>) {
    if (!findMatches.has(lv.vertex)) {
      return null;
    }
    return (
      <rect
        className="DdgNode-VectorFindEmphasis--outline"
        vectorEffect="non-scaling-stroke"
        width={lv.width - 2}
        height={lv.height - 2}
        x="1"
        y="1"
      />
    );
  }

  function renderHtmlFindEmphasis(lv: TLayoutVertex<any>) {
    if (!findMatches.has(lv.vertex)) {
      return null;
    }
    return <div className="DdgNode-HtmlFindEmphasis" />;
  }

  function renderVectorFindEmphasisColorBand(lv: TLayoutVertex<any>) {
    if (!findMatches.has(lv.vertex)) {
      return null;
    }
    return (
      <rect
        className="DdgNode-VectorFindEmphasis--colorBand"
        vectorEffect="non-scaling-stroke"
        width={lv.width - 2}
        height={lv.height - 2}
        x="1"
        y="1"
      />
    );
  }

  return {
    htmlFindEmphasis: renderHtmlFindEmphasis,
    vectorBorder: renderVectorBorder,
    vectorFindColorBand: renderVectorFindEmphasisColorBand,
    vectorFindOutline: renderVectorFindEmphasisOutline,
  };
}
