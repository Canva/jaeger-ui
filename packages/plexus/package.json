{
  "name": "@jaegertracing/plexus",
  "version": "0.0.1-dev.2",
  "description": "Direct Graph React component",
  "main": "umd/@jaegertracing/plexus.js",
  "files": ["lib", "umd"],
  "scripts": {
    "build": "nwb build-react-component",
    "clean": "nwb clean-module && nwb clean-demo",
    "start": "nwb serve-react-demo",
    "test": "echo 'NO TESTS YET'",
    "coverage": "echo 'NO TESTS YET'",
    "test:dev": "nwb test-react --server"
  },
  "dependencies": {
    "viz.js": "^1.8.1"
  },
  "peerDependencies": {
    "react": "16.x"
  },
  "devDependencies": {
    "nwb": "0.21.x",
    "react": "^16.3.2",
    "react-dom": "^16.3.2",
    "worker-loader": "^1.1.1"
  },
  "author": "",
  "homepage": "https://github.com/jaegertracing/jaeger-ui",
  "license": "Apache-2.0",
  "repository": "https://github.com/jaegertracing/jaeger-ui.git",
  "keywords": ["react-component"]
}
