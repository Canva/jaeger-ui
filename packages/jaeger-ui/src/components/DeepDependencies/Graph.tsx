// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { PureComponent } from 'react';
import { DirectedGraph, LayoutManager } from '@jaegertracing/plexus';
import { TEdge } from '@jaegertracing/plexus/lib/types';

import getNodeLabel from './getNodeLabel';

import { TDdgVertex } from '../../model/ddg/types';

type TProps = {
  edges: TEdge[];
  uiFindMatches: Set<TDdgVertex> | undefined;
  vertices: TDdgVertex[];
};

export default class Graph extends PureComponent<TProps> {
  private layoutManager: LayoutManager;

  constructor(props: TProps) {
    super(props);
    this.layoutManager = new LayoutManager({ useDotEdges: true, splines: 'polyline' });
  }

  render() {
    const { edges, uiFindMatches, vertices } = this.props;

    return (
      <DirectedGraph
        minimap
        zoom
        arrowScaleDampener={0}
        minimapClassName="u-miniMap"
        layoutManager={this.layoutManager}
        edges={edges}
        vertices={vertices}
        getNodeLabel={getNodeLabel(uiFindMatches)}
      />
    );
  }
}
