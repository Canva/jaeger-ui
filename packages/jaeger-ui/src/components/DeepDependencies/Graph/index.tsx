// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { PureComponent } from 'react';
import memoize from 'memoize-one';
import { Digraph, LayoutManager } from '@jaegertracing/plexus';
import { TSetProps, TFromGraphStateFn, TDefEntry } from '@jaegertracing/plexus/lib/Digraph/types';
import { TEdge } from '@jaegertracing/plexus/lib/types';
import TNonEmptyArray from '@jaegertracing/plexus/lib/types/TNonEmptyArray';

import DdgNodeContent from './DdgNodeContent';
import getNodeRenderers from './getNodeRenderers';
import getSetOnEdge from './getSetOnEdge';
import { PathElem, TDdgVertex, EViewModifier } from '../../../model/ddg/types';

import './index.css';

type TProps = {
  edges: TEdge[];
  edgesViewModifiers: Map<string, number>;
  getVisiblePathElems: (vertexKey: string) => PathElem[] | undefined;
  setViewModifier: (vertexKey: string, viewModifier: EViewModifier, enable: boolean) => void;
  uiFindMatches: Set<TDdgVertex> | undefined;
  vertices: TDdgVertex[];
  verticesViewModifiers: Map<string, number>;
};

const { scaleStrokeOpacityStrongest } = Digraph.propsFactories;

const setOnEdgesContainer: TSetProps<TFromGraphStateFn<TDdgVertex, any>> = [
  scaleStrokeOpacityStrongest,
  { stroke: '#444', strokeWidth: 0.7 },
];

const edgesDefs: TNonEmptyArray<TDefEntry<TDdgVertex, unknown>> = [
  { localId: 'arrow' },
  { localId: 'arrow-hovered', setOnEntry: { className: 'DdgArrow is-pathHovered' } },
];

export default class Graph extends PureComponent<TProps> {
  private getNodeRenderers = memoize(getNodeRenderers);
  private getNodeContentRenderer = memoize(DdgNodeContent.getNodeRenderer);
  private getSetOnEdge = memoize(getSetOnEdge);

  private layoutManager: LayoutManager = new LayoutManager({
    useDotEdges: true,
    splines: 'polyline',
    rankdir: 'TB',
  });

  private emptyFindSet: Set<TDdgVertex> = new Set();

  componentWillUnmount() {
    this.layoutManager.stopAndRelease();
  }

  render() {
    const {
      edges,
      edgesViewModifiers,
      getVisiblePathElems,
      setViewModifier,
      uiFindMatches,
      vertices,
      verticesViewModifiers,
    } = this.props;
    const findRenderers = this.getNodeRenderers(uiFindMatches || this.emptyFindSet, verticesViewModifiers);

    return (
      <Digraph<TDdgVertex>
        minimap
        zoom
        minimapClassName="u-miniMap"
        layoutManager={this.layoutManager}
        edges={edges}
        vertices={vertices}
        measurableNodesKey="nodes/content"
        layers={[
          {
            key: 'nodes/find-emphasis/vector-outline',
            layerType: 'svg',
            renderNode: findRenderers.vectorFindOutline,
          },
          {
            key: 'nodes/find-emphasis/html',
            layerType: 'html',
            renderNode: findRenderers.htmlFindEmphasis,
          },
          {
            key: 'nodes/find-emphasis/vector-color-band',
            layerType: 'svg',
            renderNode: findRenderers.vectorFindColorBand,
          },
          {
            key: 'nodes/vector-border',
            layerType: 'svg',
            renderNode: findRenderers.vectorBorder,
          },
          {
            key: 'edges',
            layerType: 'svg',
            edges: true,
            defs: edgesDefs,
            markerEndId: 'arrow',
            setOnContainer: setOnEdgesContainer,
            setOnEdge: this.getSetOnEdge(edgesViewModifiers),
          },
          {
            key: 'nodes/content',
            layerType: 'html',
            measurable: true,
            measureNode: DdgNodeContent.measureNode,
            renderNode: this.getNodeContentRenderer(getVisiblePathElems, setViewModifier),
          },
        ]}
      />
    );
  }
}
