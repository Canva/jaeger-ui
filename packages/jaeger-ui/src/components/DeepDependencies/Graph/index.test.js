// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { shallow } from 'enzyme';
import { Digraph, LayoutManager } from '@jaegertracing/plexus';

import Graph from './index';

describe('<Graph />', () => {
  const vertices = [...new Array(10)].map((_, i) => ({ key: `key${i}` }));
  const edges = [
    {
      from: vertices[0].key,
      to: vertices[1].key,
    },
    {
      from: vertices[1].key,
      to: vertices[2].key,
    },
  ];

  const props = {
    vertices,
    edges,
    edgesViewModifiers: new Map(),
  };

  describe('constructor', () => {
    it('creates layout manager', () => {
      const graph = new Graph(props);
      expect(graph.layoutManager instanceof LayoutManager).toBe(true);
    });
  });

  describe('render', () => {
    it('renders provided edges and vertices', () => {
      const wrapper = shallow(<Graph {...props} />);
      const plexusGraph = wrapper.find(Digraph);
      expect(plexusGraph.prop('edges')).toEqual(edges);
      expect(plexusGraph.prop('vertices')).toEqual(vertices);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
