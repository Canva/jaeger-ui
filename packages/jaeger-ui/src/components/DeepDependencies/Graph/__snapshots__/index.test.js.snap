// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Graph /> render renders provided edges and vertices 1`] = `
<Digraph
  className=""
  classNamePrefix="plexus"
  edges={
    Array [
      Object {
        "from": "key0",
        "to": "key1",
      },
      Object {
        "from": "key1",
        "to": "key2",
      },
    ]
  }
  layers={
    Array [
      Object {
        "key": "nodes/find-emphasis/vector-outline",
        "layerType": "svg",
        "renderNode": [Function],
      },
      Object {
        "key": "nodes/find-emphasis/html",
        "layerType": "html",
        "renderNode": [Function],
      },
      Object {
        "key": "nodes/find-emphasis/vector-color-band",
        "layerType": "svg",
        "renderNode": [Function],
      },
      Object {
        "key": "nodes/vector-border",
        "layerType": "svg",
        "renderNode": [Function],
      },
      Object {
        "defs": Array [
          Object {
            "localId": "arrow",
          },
          Object {
            "localId": "arrow-hovered",
            "setOnEntry": Object {
              "className": "DdgArrow is-pathHovered",
            },
          },
        ],
        "edges": true,
        "key": "edges",
        "layerType": "svg",
        "markerEndId": "arrow",
        "setOnContainer": Array [
          [Function],
          Object {
            "stroke": "#444",
            "strokeWidth": 0.7,
          },
        ],
        "setOnEdge": undefined,
      },
      Object {
        "key": "nodes/content",
        "layerType": "html",
        "measurable": true,
        "measureNode": [Function],
        "renderNode": [Function],
      },
    ]
  }
  layoutManager={
    LayoutManager {
      "_handleUpdate": [Function],
      "coordinator": Coordinator {
        "_handleVizWorkerError": [Function],
        "_handleVizWorkerMessage": [Function],
        "_handleVizWorkerMessageError": [Function],
        "busyWorkers": Array [],
        "callback": [Function],
        "currentLayout": null,
        "idleWorkers": Array [],
        "nextWorkerId": 0,
      },
      "layoutId": 0,
      "options": Object {
        "rankdir": "TB",
        "splines": "polyline",
        "useDotEdges": true,
      },
      "pendingResult": null,
    }
  }
  measurableNodesKey="nodes/content"
  minimap={true}
  minimapClassName="u-miniMap"
  vertices={
    Array [
      Object {
        "key": "key0",
      },
      Object {
        "key": "key1",
      },
      Object {
        "key": "key2",
      },
      Object {
        "key": "key3",
      },
      Object {
        "key": "key4",
      },
      Object {
        "key": "key5",
      },
      Object {
        "key": "key6",
      },
      Object {
        "key": "key7",
      },
      Object {
        "key": "key8",
      },
      Object {
        "key": "key9",
      },
    ]
  }
  zoom={true}
/>
`;
